### ISSUE TITLE
parking_lot_core 0.8.1 wasm link issue

### ISSUE THREAD

**Issue by emilk** [Amanieu/parking_lot#269]
Starting with parking_lot_core 0.8.1, when compiling to wasm there is now a depdendency on "env".

This means one can no longer use parking_lot from within a wasm runtime, such as [`wasmtime`](https://github.com/bytecodealliance/wasmtime).

When compiling for the web, one can address this problem with `parking_lot = { version = "0.11", features = ["wasm-bindgen"] }`, but this does not work for non-web wasm runtimes.

## Reproduce

`Cargo.toml`:

``` toml
[package]
name = "parking_lot_test"
version = "0.1.0"

[dependencies]
parking_lot = "0.11"
parking_lot_core = "=0.8.1"
```

`main.rs`:

``` rust
fn main() {
    let m = parking_lot::Mutex::new(());
    let _lock = m.lock();
}
```

Using [`wasm2wat`](https://github.com/WebAssembly/wabt):

```
§ cargo build --target "wasm32-unknown-unknown"
§ wasm2wat target/wasm32-unknown-unknown/debug/parking_lot_test.wasm | rg import
> (import "env" "now" (func $now (type 7)))
```

(with `parking_lot_core = "=0.8.0"` there is no `import "env"`)

**Comment by bjorn3** [comment]
https://github.com/Amanieu/parking_lot/compare/0.8.0...0.8.1?w=1#diff-f8a0a8b9353ddc2e1efd135d75ec0668f64e2674c15e9b1cda089cbd69f60372 shows no difference in the parking implementation between v0.8.0 and v0.8.1 that could explain this. Does this still happen with LTO enabled?

**Comment by emilk** [comment]
I tested it now with `--release` and `[profile.release] lto = true` – same problem.

**Comment by emilk** [comment]
I suspect this is related to `parking_lot_core`:s use of https://github.com/sebcrozet/instant

**Comment by bjorn3** [comment]
The parking_lot algorithm itself need to know the current time for fair locks but the single-threaded wasm option shouldn't need this as there are no other threads against which it needs to be fair.

**Comment by emilk** [comment]
Interesting!

Is there a way to tell parking_lot that it is compiling for a single-threaded target, so that it does not require the `instant` crate at all? If not, should there be?

**Comment by bjorn3** [comment]
It should compile for a single-threaded target by default. Multi-threading requires the bulk memory wasm proposal and atomic instructions. Both of which are not supported for the WASM MVP that is targeted by default. In addition it requires all parts of the program to be compiled with multi-threading support I believe, but libstd isn't by default.

**Comment by emilk** [comment]
I suspect that the only reason parking_lot 0.8.0 worked and 0.8.1 doesn't is because some subtle failure of rustc to elide unused code (using instant).

---

For my use-case ("pure wasm" = no instant) we would need cfg-blocks around all uses of the `instant` crate, as well as around functions that uses time, such as `wait_until`, `try_lock_until` etc.

For example:

``` rust
#[cfg(feature = "instant")]
unsafe fn park_until(&self, timeout: Instant) -> bool { ... }
````

This would be a very intrusive change to the parking_lot code base, disabling all uses of times (`Instant` and `Duration`).

(Using `#[cfg(not(target_arch = "wasm32"))]` would not be a good idea, because when compiling wasm to the web `instant` works fine with `features = ["wasm-bindgen"]`).

--- 

Another alternative is that I create a new wrapper crate around parking_lot which exposes only what I need (`Mutex` and `RwLock`) and that has a feature-flag for single-threaded, no-instant compile-targets that just uses a `bool` to check if a `Mutex` is already locked (just to panic on reentrant double-lock).

**Comment by faern** [comment]
Why do you need locking primitives in a single threaded environment?

**Comment by emilk** [comment]
> Why do you need locking primitives in a single threaded environment?

Great question!

Say you have a library that may or may not be used from a multithreaded context, so the library uses `parking_lot::Mutex` in a few places. There is also an app that wants to use that library. The app happens to be compiled to a singlethreaded wasm32 target. _Should this be possible?_

The answer is: it used to be possible, but with parking_lot_core 0.8.1 it no longer is.

So who's responsibility is it to fix this?

The app can do a work-around by adding`parking_lot_core = "=0.8.1"` to it's Cargo.toml, but that work-around will only work for so long. As soon as the library upgrades to a future `parking_lot_core = 0.9.0" then the app can no use the latest version of that library.

The library could add feature flags around all uses of `parking_lot`, but there are a lot of libraries using `parking_lot`, and the app developer will have trouble changing all the libraries.

The third place to fix this is in parking_lot itself, but saying "parking_lot supports pure wasm32 contexts". I think this is the route that is the best.

---

In summary:

Either we say "parking_lot should work for pure wasm32 targets" and work towards that, **OR** we say "if you use `parking_lot` in your library, then your library cannot be used by apps that want to compile to pure wasm32 targets".

**Comment by faern** [comment]
That makes sense. I got the imoression that your code needed it directly and your code was targeting single threaded wasm only :)

**Comment by ibaryshnikov** [comment]
Isn't it related to this?
https://github.com/rust-lang/rust/issues/72758#issuecomment-636184695

`(import "env" "now" (func $now (type 7)))`
it also feels like coming from this line
https://github.com/Amanieu/parking_lot/compare/0.8.0...0.8.1#diff-c6e37a4985a3b050dda1f692981c6de9f3d039f86312603801f9901167e843a8R57

```rust
if let Some(left) = timeout.checked_duration_since(Instant::now()) {
```

**Comment by bjorn3** [comment]
That one is only used when wasm atomics are enabled, which aren't by default.

**Comment by ibaryshnikov** [comment]
Sorry, I misread some of the comments, it feels like the source and the solution have already been found.

**Comment by repi** [comment]
Would indeed be great to clarify if non-web `wasm32-unknown-unknown` is intended to not be supported anymore with parking_lot, or if it is and then fix this by not depending on `instant` and disable the parts of the API that require it timing access.

**Comment by Amanieu** [comment]
`wasm32-unknown-unknown` is a bit of weird target: it really should have been a `#[no_std]` target but instead comes with a mostly non-functional `std`. `parking_lot` does require a working `std` for various features such as `Instant` (for timed locks/waits), thread-local storage, blocking in the OS, etc.

I'm happy to accept any PRs to support `wasm32-unknown-unknown` (e.g. by simply disabling all time-related functionality) but in the end this target will have to be supported by users and not by myself (since I don't use WASM personally and am not familiar with all the ways it can be embedded).

**Comment by repi** [comment]
Yeah `wasm32-unknown-unknown` is indeed an odd target, would have preferred it to be a lean no-std target (which is how we use it) and have the web bindings using a separate `wasm32-web` (or similar target), and WASI for std-target for running primarily outside of browser.

For now we haven't run into that many dependencies that we use in `wasm32-unknown-unknown` that also uses `parking_lot`, mainly a few of our own that uses `parking_lot` instead of std as faster primitives for native compilation and for wasm intending to have it just be no-ops in most cases as it is a singlethreaded no-std environment. But so far we've been able to just remove the `parking_lot` dependencies, that could get harder going forward though but works for now.

**Comment by fosskers** [comment]
This recently popped up again, as https://lib.rs/crates/string_cache just got an update, adding `parking_lot` as a dependency. `string_cache` in turn is a dependency of https://lib.rs/crates/scraper .

**Comment by dsherret** [comment]
My temporary workaround is to add the following to my Cargo.toml file.

```toml
[target.'cfg(target_arch = "wasm32")'.dependencies]
parking_lot_core = "=0.8.0"
```

**Comment by d3lm** [comment]
@Amanieu 🙌
